chore: CAP (#59)
katcipis/sophia: notes/books/designing-data-intensive-apps.md

provides. In a non-linearizable database you would need to design the solution differently or
implement some domain specific compare-and-set strategy, like receiving a hash of the image on the
message and if the hash doesn't match fail (with some retry strategy).

## CAP Theorem

There is some interesting remarks on the limitations of the CAP theorem. It was useful as a tool
to start discussions a decade ago but the author defends that it is an outdated idea that
is not very useful when discussing trade-offs in distributed systems. Some limitations include:

1 - Easilly misunderstood as a binary trade-off, like you lose availabiliy to get consistency and vice
versa. Trade-offs in distributed systems are much more nuanced and complex than this.

2 - Introduces partitioning as a variable on the equation, which on distributed systems doesn't apply since
it would imply single node systems (to avoid partitioning). So in the end most discussions end up being
about consistency VS availability.

3 - Even thinking on terms of consistency VS availability doesn't provide a complete model.
It ignores another major architectural property that must be taken in consideration: performance.

> One of the points that made me think the most is performance because there is an concrete example
> that makes it very obvious the limitations of CAP when thinking about consistency models, the memory
> models of multi-core SMP CPUs.

Reviewer: That grammar here is not quite right. It should be:

```suggestion
One of the points that made me think the most is performance because there is a concrete example
that makes the limitations of CAP when thinking about consistency models very obvious, the memory
models of multi-core SMP CPUs.
```

Or maybe this:

> One of the points that made me think the most is performance because there is a concrete example
that really shows the limitations of CAP when thinking about consistency models, the memory
models of multi-core SMP CPUs.

considerably inconsistent. And yet there is no partitioning whatsoever in this case, the CPU cores exist
on the same Die and there is no way they will get disconnected/partitioned (unless the CPU is broken,
then it is a total fault).

The same applies for memory, memory is local. All communication can be synchronous and have a well defined and
deterministic upper bound on worse case delays, you literally know the physical distance between components.
This fits perfectly the theoretical definition of the synchronous model, you know exactly how long each
communication can take and you can be sure if the communication failed or not (no uncertainty).

And yet, communication between different cores and their caches is done asynchronously. Why ? Because of
performance. When a CPU core writes in memory, in order to guarantee linearizability it would have to synchronize
with all other CPU cores and their caches, reducing paralelism considerably and complicating CPU design
(or write in memory + invalidate the cache of all other cores, which would make caches nearly useless).

The whole idea of linearizability is for data to behave like only a single copy of it exists and all

> operatios are atomic. In a SMP multi-core CPU this costs too much, the performance gains from each core having

Reviewer: ```suggestion
operatios are atomic. In an SMP multi-core CPU this costs too much, the performance gains from each core having
```

SMP starts with a vowel -- at least when you say it out loud -- so it should be "an".

Author: I'm quite bad with "a" vs "an" TBH, so keep it coming :rofl:

Reviewer: Oh, what I meant was if you say SMP out loud, it sounds like "ess em pee". So the way it's written it starts with an S, but the way you say it it starts with an E.

Author: Ah got it :100:
